### ISSUE TITLE
Things are being initialised before XML files are loaded

### ISSUE THREAD

**Issue by cdjackson** [eclipse/smarthome#1143]
The log below shows that things can be initialised before the XML files are loaded. This causes problems in ZWave at least as once the protocol side of the binding starts up, it reads through the things to look for the thing type definition that's related to the Z-Wave device (using properties that is stored in the XML).

I think there needs to be some way to either ensure the XML is loaded before initialising the bindings things, or there needs to be a callback after the XML is loaded? Or maybe there's another way to know, or force the XML to be loaded from within the binding?

```
2016-03-06 19:24:03.348 [DEBUG] [.b.z.handler.ZWaveThingHandler:136  ] - NODE 14: Initialising channel zwave:everspring_st814_00_000:1530ee834f6:node14:sensor_relhumidity2
2016-03-06 19:24:03.349 [DEBUG] [o.b.z.h.ZWaveControllerHandler:70   ] - Initializing ZWave Controller.
2016-03-06 19:24:03.353 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_dsd37_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.350 [DEBUG] [.b.z.handler.ZWaveThingHandler:276  ] - NODE 14: Controller initialised. Starting device intialisation.
2016-03-06 19:24:03.357 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw062_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.371 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:everspring_st814_00_000:1530ee834f6:node14' updated: OFFLINE
2016-03-06 19:24:03.371 [DEBUG] [.b.z.handler.ZWaveThingHandler:276  ] - NODE 14: Controller initialised. Starting device intialisation.
2016-03-06 19:24:03.372 [DEBUG] [.b.z.handler.ZWaveThingHandler:276  ] - NODE 44: Controller initialised. Starting device intialisation.
2016-03-06 19:24:03.372 [WARN ] [.b.z.handler.ZWaveThingHandler:306  ] - NODE 14: Controller failed to register event handler.
2016-03-06 19:24:03.372 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:serial_zstick:1530ee834f6' updated: ONLINE
2016-03-06 19:24:03.371 [DEBUG] [.b.z.handler.ZWaveThingHandler:276  ] - NODE 14: Controller initialised. Starting device intialisation.
2016-03-06 19:24:03.372 [WARN ] [.b.z.handler.ZWaveThingHandler:306  ] - NODE 44: Controller failed to register event handler.
2016-03-06 19:24:03.372 [INFO ] [b.z.handler.ZWaveSerialHandler:77   ] - Connecting to serial port '/dev/tty.SLAB_USBtoUART'
2016-03-06 19:24:03.372 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'zwave:everspring_st814_00_000:1530ee834f6:node14' changed from INITIALIZING to OFFLINE
2016-03-06 19:24:03.373 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'zwave:serial_zstick:1530ee834f6' changed from INITIALIZING to ONLINE
2016-03-06 19:24:03.373 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/binding/binding.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.373 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:everspring_st814_00_000:1530ee834f6:node14' updated: OFFLINE
2016-03-06 19:24:03.374 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:popp_123665_00_000:1530ee834f6:node44' updated: OFFLINE
2016-03-06 19:24:03.374 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:everspring_st814_00_000:1530ee834f6:node14' updated: OFFLINE
2016-03-06 19:24:03.372 [WARN ] [.b.z.handler.ZWaveThingHandler:306  ] - NODE 14: Controller failed to register event handler.
2016-03-06 19:24:03.376 [DEBUG] [c.x.o.XmlDocumentBundleTracker:159  ] - Create an empty XmlDocumentProvider for the module 'org.openhab.binding.zwave'.
2016-03-06 19:24:03.382 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw074_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.375 [DEBUG] [.c.t.i.TransformationActivator:34   ] - Transformation Service has been started.
2016-03-06 19:24:03.384 [DEBUG] [.b.z.handler.ZWaveThingHandler:271  ] - NODE 14: Polling intialised at 1800 seconds.
2016-03-06 19:24:03.392 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'zwave:serial_zstick:1530ee834f6' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Serial Error: Port /dev/tty.SLAB_USBtoUART does not exist
2016-03-06 19:24:03.393 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:everspring_st814_00_000:1530ee834f6:node14' updated: OFFLINE (BRIDGE_OFFLINE)
2016-03-06 19:24:03.393 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw075_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.393 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'zwave:popp_123665_00_000:1530ee834f6:node44' updated: OFFLINE (BRIDGE_OFFLINE)
2016-03-06 19:24:03.393 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'zwave:everspring_st814_00_000:1530ee834f6:node14' changed from OFFLINE to OFFLINE (BRIDGE_OFFLINE)
2016-03-06 19:24:03.394 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'zwave:popp_123665_00_000:1530ee834f6:node44' changed from OFFLINE to OFFLINE (BRIDGE_OFFLINE)
2016-03-06 19:24:03.404 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw089_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.410 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw096_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.421 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw097_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.424 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw098_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.431 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/aeon_zw100_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.439 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/benext_plugindimmer_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.453 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/brk_zcombo_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.456 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/cooper_rf9500_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.466 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/cooper_rftr9505_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.468 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/danfoss_lc13_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.469 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/danfoss_lcz251_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-03-06 19:24:03.472 [DEBUG] [c.x.o.XmlDocumentBundleTracker:118  ] - Reading the XML document '/ESH-INF/thing/danfoss_mt02650_0_0.xml' in module 'org.openhab.binding.zwave'...
```

**Comment by cdjackson** [comment]
I was wondering if there was any thoughts on how to solve this.  This issue is getting worse for Z-Wave - we now have 196 things in the binding with around 330 in the database at the moment...

Is it something that is planned to be fixed (I see you added the 'bug' label) or do I need to try and find a workaround? It takes around 4 seconds to load the XML files on my Mac - I don't think it's a big issue if it takes 30 seconds or more - it's a one off startup event, but the binding needs to be able to work with this delay and I need to find some way to delay startup until the XML is available.

There are two issues I have with this -:

* Shortly after the bridge and thing handler are initialized, I do some updates on the configuration to synchronise the information on the device, with the information held in the ESH config. This is just needed to make sure that the user has the current view of the configuration. Currently, the configuration description isn't always loaded when the thing handler starts.
* When a new thing is added, I search through all the XML files to find one that meets the device information. If discovery is started automatically when the binding starts (as it is in Z-Wave) then sometimes the device isn't found as the XML isn't available.

I might be able to find workarounds to this if this isn't going to be fixed in the near future (although I've no idea at the moment what the workaround might be...).

**Comment by kaikreuzer** [comment]
I added the "bug" label because I think it is a serious flaw in the current core architecture and thus needs to be addressed there (so my clear answer is that I do NOT expect you having to workaround it).

I am not yet clear on how to tackle the problem best, though. Loading 4 secs on a Mac means that it already now takes maybe 20-30 seconds on a RaspPi - this doesn't sound good to me.

I remember your suggestion from a long time ago that we sooner or later will need a kind of index for a quick lookup of the most important infos - really just to be able to provide a list of thing types, without having to provide them in all their glory. I think this is probably the only way forward that will not only postpone the problem. Such an index could be read upon bundle start, so that there are no async processing issues as the one discussed here.

@sjka What is your suggestion on how to proceed here?

**Comment by cdjackson** [comment]
> I am not yet clear on how to tackle the problem best, though. Loading 4 secs on a Mac means that it already now takes maybe 20-30 seconds on a RaspPi - this doesn't sound good to me.
> 
I agree it’s not optimum, but on the other hand, does it really matter? Hopefully people aren’t restarting the system often, so it shouldn’t be a major problem (IMHO).


> I remember your suggestion from a long time ago that we sooner or later will need a kind of index for a quick lookup of the most important infos 
> 
Yes - I didn’t want to add that in here just yet as I didn’t think it solved the problem, but maybe it’s at least part of the solution. The problem would still be one of the previous examples where I would like to have access to the config descriptions when the handler starts. It might be acceptable to get some sort of notification once the xml for this thing is loaded if it’s after the hander has started, but the index doesn’t directly solve this.

**Comment by maggu2810** [comment]
If I understand the whole problem correctly it seems to me a must, that the core uses a binding not before all files in ESH-INF has been parsed.
Without respect how this could be solved:
A binding should rely on the fact, that all stuff it provides in the ESH-INF directory is available if its logic is used.

Is this what do you want @cdjackson?

**Comment by sjka** [comment]
> Without respect how this could be solved

Well, that heavily depends on what we define to be "the logic to be used" is. Primarily that would be the ```ThingHandler.initialize()``` call, I would assume. We could defer that one, I think, in one way or the other. The initialize() is called asynchronously anyway, so it should not matter to store it in a queue and wait until everything has been loaded...
Regarding the "loading finished" detection, we have the XmlDocumentBundleTracker. We could extend it in a way so that we can ask it for any given bundle what the loading state is. It would be finished only once the XmlDocumentBundleTracker knows that the bundle either has no XMLs, or the existing ones have been loaded completely. 

Would that help?

**Comment by cdjackson** [comment]
>A binding should rely on the fact, that all stuff it provides in the ESH-INF directory is available if its logic is used.

> Is this what do you want @cdjackson?

Yes - to me this is the requirement
